serverless/cmd/feeder/impl/feeder.go

```diff
@@ -0,0 +1,161 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// impl is a witness feeder implementation for the serverless log.
+package impl
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/client"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// Witness describes the operations the feeder needs to interact with a witness.
+type Witness interface {
+	// SigVerifier returns a verifier which can check signatures from this witness.
+	SigVerifier() note.Verifier
+	// GetLatestCheckpoint returns the latest checkpoint the witness holds for the given logID.
+	// Must return os.ErrNotExists if the logID is know, but it has no checkpoint for that log.
+	GetLatestCheckpoint(ctx context.Context, logID string) ([]byte, error)
+	// Update attempts to clock the witness forward for the given logID.
+	Update(ctx context.Context, logID string, newCP []byte, proof [][]byte) error
+}
+
+// FeedOpts holds parameters when calling the Feed function.
+type FeedOpts struct {
+	// LogID is the ID for the log whose checkpoint is being fed.
+	LogID string
+	// LogFetcher for the source log (used to build proofs).
+	LogFetcher client.Fetcher
+	// LogSigVerifier a verifier for log checkpoint signatures.
+	LogSigVerifier note.Verifier
+
+	// NumRequired is the number of witness signatures required for a call to Feed to be considered successful.
+	NumRequired int
+
+	// Witnesses is a list of witnesses to feed to.
+	Witnesses []Witness
+}
+
+// Feed sends the provided checkpoint to the configured set of witnesses.
+// Returns the provided checkpoint plus at least cfg.NumRequired signatures.
+func Feed(ctx context.Context, cp []byte, opts FeedOpts) ([]byte, error) {
+	if nw := len(opts.Witnesses); opts.NumRequired > nw {
+		return nil, fmt.Errorf("NumRequired is %d, but only %d witnesses provided", opts.NumRequired, nw)
+	}
+	n, err := note.Open(cp, note.VerifierList(opts.LogSigVerifier))
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify signature on checkpoint: %v", err)
+	}
+	cpSubmit := &log.Checkpoint{}
+	_, err = cpSubmit.Unmarshal([]byte(n.Text))
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify signature on checkpoint to witness: %v", err)
+	}
+
+	sigs := make(chan note.Signature, len(opts.Witnesses))
+	errs := make(chan error, len(opts.Witnesses))
+	// TODO(al): make this configurable
+	h := hasher.DefaultHasher
+
+	for _, w := range opts.Witnesses {
+		go func(ctx context.Context, w Witness) {
+			// Keep submitting until success or context timeout...
+			t := time.NewTicker(1)
+			wSigV := w.SigVerifier()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-t.C:
+					t.Reset(5 * time.Second)
+				}
```

Comment:
But then I'd have to build large trees of if/else in order to hit it rather than popping back to the top with the `continue`s, wouldn't I? (I guess I could use a `goto` if you want?)

Reply:
Yeah that's true. Any other changes would require a lot of refactoring so let's leave it.